// next/dist/esm/server/render-result.js
export default class RenderResult {
    constructor(response, { contentType , ...metadata } = {}){
        this._response = response;
        this._contentType = contentType;
        this._metadata = metadata;
    }
    get metadata() {
        return this._metadata;
    }
    isNull() {
        return this._response === null;
    }
    contentType() {
        return this._contentType;
    }
    toUnchunkedString() {
        if (typeof this._response !== "string") {
            throw new Error("invariant: dynamic responses cannot be unchunked. This is a bug in Next.js");
        }
        return this._response;
    }
    pipe(res) {
        if (this._response === null) {
            throw new Error("invariant: response is null. This is a bug in Next.js");
        }
        if (typeof this._response === "string") {
            throw new Error("invariant: static responses cannot be piped. This is a bug in Next.js");
        }
        const response = this._response;
        const flush = typeof res.flush === "function" ? ()=>res.flush() : ()=>{};
        return (async ()=>{
            const reader = response.getReader();
            let fatalError = false;
            try {
                while(true){
                    const { done , value  } = await reader.read();
                    if (done) {
                        res.end();
                        return;
                    }
                    fatalError = true;
                    res.write(value);
                    flush();
                }
            } catch (err) {
                if (fatalError) {
                    res.destroy(err);
                }
                throw err;
            }
        })();
    }
    isDynamic() {
        return typeof this._response !== "string";
    }
    static fromStatic(value) {
        return new RenderResult(value);
    }
    static #_ = (()=>{
        this.empty = RenderResult.fromStatic("");
    })();
}
